File: p007.py
#!/usr/bin/python


def nth_prime(n, limit=125000):
    if limit % 2 != 0:
        limit += 1
    primes = [True] * limit
    primes[0], primes[1] = [None] * 2
    cnt = 0
    for ind, val in enumerate(primes):
        if val is True:
            # sieve out non-primes by multiples of known primes
            primes[ind * 2::ind] = [False] * (((limit - 1) // ind) - 1)
            cnt += 1
        if cnt == n:
            return ind
    return False


print(nth_prime(10001))

File: rust/src/bin/p007.rs

fn nth_prime(n: usize) -> Option<usize> {
    let mut primes = [true; 125000];
    primes[0] = false;
    primes[1] = false;
    let (mut cnt, mut ind) = (0, 0);
    while ind < primes.len() {
        if primes[ind] == true {
            let mut i = ind * ind;
            while i < primes.len() {
                primes[i] = false;
                i += ind;
            }
            cnt += 1;
        }
        if cnt == n {
            return Some(ind)
        }
        ind += 1;
    }
    None
}

fn main() {
    println!("{}", nth_prime(10001).unwrap());
}
